onnxruntime/core/graph/function_utils.cc

```diff
@@ -0,0 +1,672 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/graph/function_utils.h"
+#include "core/common/inlined_containers.h"
+#include "core/framework/tensorprotoutils.h"
+#include "onnx/shape_inference/implementation.h"
+#include "core/graph/function_impl.h"
+#include "core/graph/model_load_utils.h"
+
+namespace ONNX_NAMESPACE {
+// Infer shape for functions. This also supports
+// nested model local functions.
+// TODO: Add this to onnx instead of adding it here.
+void InferShapeForFunctionNode(
+    InferenceContext& ctx,
+    const FunctionProto& func_proto,
+    const std::unordered_map<std::string, int>& func_opset_imports,
+    const ShapeInferenceOptions& options,
+    const ISchemaRegistry* schema_registry,
+    const std::unordered_map<std::string, const FunctionProto*>& in_model_functions,
+    std::function<std::string(const std::string& function_domain, const std::string& function_name)> get_func_id) {
+  GraphProto g;
+  // Get a temporary tensor-shape map
+  const auto num_func_inputs = func_proto.input_size();
+  std::unordered_map<std::string, TypeProto*> value_types_by_name;
```

Comment:
>std::unordered_map<std::string,  [](http://example.com/codeflow?start=2&length=32)

InlinedHashMap

Reply:
I have applied the inlined container to most places, but for the maps used in this "[InferShapeForFunctionNode](https://github.com/microsoft/onnxruntime/blob/chenta/function_refactory/onnxruntime/core/graph/function_utils.cc#L14)", it is not that easy. The reason i have to keep it with std::unordered_map is because those codes take the dependency to the signature of onnx shape inference (see [here](https://github.com/microsoft/onnxruntime/blob/chenta/function_refactory/onnxruntime/core/graph/function_utils.cc#L14) and [here](https://github.com/onnx/onnx/blob/main/onnx/shape_inference/implementation.h#L128), so i leave those function untouched. 

Those codes are not new codes introduced, they already exist in our codebase, this PR just move it to a new place, so even i left it as std::unordered_map, it won't be worse than what we have now.

Reply:
I would also focus on eliminating the duplication between this and ONNX inference code first. We may not need to have this code exist in onnxruntime.